tools/github/release/manager.py

```diff
@@ -0,0 +1,143 @@
+import pathlib
+import re
+from functools import cached_property
+from typing import Dict, List, Optional, Pattern, Type, Union
+
+import verboselogs  # type:ignore
+
+import packaging.version
+
+import aiohttp
+
+import gidgethub.abc
+import gidgethub.aiohttp
+
+from tools.base import abstract
+from tools.base.functional import async_property
+
+from tools.github.release.abstract import AGithubRelease, AGithubReleaseManager
+from tools.github.release.exceptions import GithubReleaseError
+# from tools.github.release.release import GithubRelease
+
+VERSION_MIN = packaging.version.Version("0")
+
+
+@abstract.implementer(AGithubReleaseManager)
+class GithubReleaseManager:
+
+    _version_re = r"v(\w+)"
+    _version_format = "v{version}"
+
+    def __init__(
+            self,
+            path: Union[str, pathlib.Path],
+            repository: str,
+            continues: Optional[bool] = False,
+            create: Optional[bool] = True,
+            user: Optional[str] = None,
+            oauth_token: Optional[str] = None,
+            version: Optional[str] = None,
+            log: Optional[verboselogs.VerboseLogger] = None,
+            asset_types: Optional[Dict[str, Pattern[str]]] = None,
+            github: Optional[gidgethub.abc.GitHubAPI] = None,
+            session: Optional[aiohttp.ClientSession] = None) -> None:
+        self.version = version
+        self._path = path
+        self.repository = repository
+        self.continues = continues
+        self._log = log
+        self.oauth_token = oauth_token
+        self.user = user
+        self._asset_types = asset_types
+        self._github = github
+        self._session = session
+        self.create = create
+
+    async def __aenter__(self) -> AGithubReleaseManager:
+        return self
+
+    async def __aexit__(self, *args) -> None:
+        await self.close()
+
+    def __getitem__(self, version) -> AGithubRelease:
+        # return self.release_class(self, version)
+        raise NotImplementedError
+
+    @property
+    def release_class(self) -> Type[AGithubRelease]:
+        # return GithubRelease
+        raise NotImplementedError
+
+    @cached_property
+    def github(self) -> gidgethub.abc.GitHubAPI:
+        return (
+            self._github
+            or gidgethub.aiohttp.GitHubAPI(self.session, self.user, oauth_token=self.oauth_token))
+
+    @cached_property
+    def log(self) -> verboselogs.VerboseLogger:
+        return self._log or verboselogs.VerboseLogger(__name__)
+
+    @cached_property
+    def path(self) -> pathlib.Path:
+        return pathlib.Path(self._path)
+
+    @async_property
+    async def latest(self) -> Dict[str, packaging.version.Version]:
```

Comment:
Is this entire class and manager just a very elaborate way to get at latest version? The contrast is some of what we do in https://github.com/envoyproxy/envoy/blob/main/tools/dependency/release_dates.py. That's also not very concise.

I thought based on the comments in the abstract interface, this would actually be doing some artifact fetch/push.

Reply:
This class does quite a lot

Overall the tool is structured broadly into 3 classes 

`GithubRelease` - CRUD operations
`GithubReleaseAssets` - upload/download of assets

and this one `GithubReleaseManager` - which:

- provides logging and warn/error handling
- holds the github connection and http client session
- iterates releases and creates a dictionary of latest major/minor versions
- version formatting/parsing (ie `v0.0.1` -> `packaging.Version("0.0.1")`)
- provides a factory for `GithubRelease` objects

its my intention to rewrite the release dates util at some point to make use of a `Runner`, improve logging/error handling, add tests etc. In that event the release dates util can almost certainly make use of this lib

> I thought based on the comments in the abstract interface, this would actually be doing some artifact fetch/push.

if you look through the interfaces again, i think you will see that this is an implementation of the manager class specified there

artifact fetch/push comes after